File: Project2_Java/src/pl/pg/oop/project2/organisms/animals/Human.java
package pl.pg.oop.project2.organisms.animals;

import pl.pg.oop.project2.organisms.Animal;
import pl.pg.oop.project2.organisms.OrganismsEnum;
import pl.pg.oop.project2.organisms.World;
import pl.pg.oop.project2.organisms.move;

import java.awt.*;

public class Human extends Animal {
    private final static int DEFAULT_INITIATIVE = 4;
    private int DEFAULT_STRENGTH = 5;
    private boolean IsAbilityOn;
    private int NumberOfRoundsUntilPossibleActivation = 0; // needs to be saved
    private final static int NumberOfTurnsWithAbilityON = 5;

    private move CurrentMove = move.NOMOVE;
    public Human(World world, Point position) {
        super(world, position);
        this.TypeOfOrganism = OrganismsEnum.HUMAN;
        this.initiative = DEFAULT_INITIATIVE;
        this.strength = DEFAULT_STRENGTH;
        this.IsAbilityOn = false;
    }

    public Human(World world, Point position, int strength) {
        super(world, position, strength);
        this.TypeOfOrganism = OrganismsEnum.HUMAN;
        this.initiative = DEFAULT_INITIATIVE;
        this.IsAbilityOn = false;
    }

    public Human(World world, Point position, int strength, int DefaultStrength, boolean IsAbilityOn, int nUntilActivation ) {
        super(world, position, strength);
        this.DEFAULT_STRENGTH = DefaultStrength;
        this.TypeOfOrganism = OrganismsEnum.HUMAN;
        this.initiative = DEFAULT_INITIATIVE;
        this.IsAbilityOn = IsAbilityOn;
        this.NumberOfRoundsUntilPossibleActivation = nUntilActivation;
    }

    /**
     * Try to turn on the Special Ability.
     * @return message about the Special Ability.
     */
    public String SpecialAbility() {
        String message;
        if (NumberOfRoundsUntilPossibleActivation == 0 && !IsAbilityOn) {
            DEFAULT_STRENGTH = strength;
            strength += NumberOfTurnsWithAbilityON;
            IsAbilityOn = true;
            message = "Special Ability Activated. Current strength: "+strength
                    +".\nWill decrease until strength is back to normal.";
        }
        else if(NumberOfRoundsUntilPossibleActivation != 0) {
            message = "Cannot activate Special Ability.\nWill be possible in: "
                    +NumberOfRoundsUntilPossibleActivation+" turns.";
        }
        else {
            message = "Ability is already on. Current strength: "+strength+".";
        }
        return message;
    }

    /**
     * Decrease the strength added by Ability.
     */
    private void DecreaseAbility() {
        if (IsAbilityOn) {
            strength -= 1;
            if (strength <= DEFAULT_STRENGTH) {
                NumberOfRoundsUntilPossibleActivation = 5;
                IsAbilityOn = false;
            }
        }
    }

    /**
     * Decrease the counter that informs if the Ability can be activated.
     */
    private void DecreaseAbilityCounter() {
        if (NumberOfRoundsUntilPossibleActivation != 0) NumberOfRoundsUntilPossibleActivation--;
    }

    /**
     * Handling Special Ability - decrease Ability or decrease the counter that informs if the Ability can be activated.
     */
    public void handleSpecialAbility() {
        DecreaseAbility();
        DecreaseAbilityCounter();
    }

    /**
     * Change Current move - human can move left, right, up, down or make no move.
     * @param NewMove The move that human will make.
     */
    public void ChangeMove(move NewMove) {
        CurrentMove = NewMove;
    }

    /**
     * @return New Position for Human based on CurrentMove.
     */
    private Point getNewHumanPosition() {
        Point newPosition;
        if (CurrentMove == move.LEFT && position.x >= step) { // Direction Left
            newPosition = new Point(position.x - step, position.y);
        }
        else if (CurrentMove == move.RIGHT && position.x < world.getSizeOfWorld().x - step) { // Direction Right
            newPosition = new Point(position.x + step, position.y);
        }
        else if (CurrentMove == move.UP && position.y >= step) { // Direction Up
            newPosition = new Point(position.x, position.y - step);
        }
        else if(CurrentMove == move.DOWN && position.y < world.getSizeOfWorld().y - step) { // Direction Down
            newPosition = new Point(position.x, position.y + step);
        }
        else {
            return new Point(-1, -1);
        }
        return newPosition;
    }

    @Override
    public boolean action() {
        Point newPosition = getNewHumanPosition();
        if(newPosition.x == -1) return false;
        prevPosition = position;
        position = newPosition;
        TryToGoToPosition();
        return true;
    }

    @Override
    public String toString() {
        return "Human";
    }

    @Override
    public String SaveOrganism() {
        return this.TypeOfOrganism + " " + this.position.x + " " + this.position.y + " "
                + this.prevPosition.x + " " + this.prevPosition.y + " " + this.strength + " "
                + this.DEFAULT_STRENGTH + " " + this.IsAbilityOn + " "
                + this.NumberOfRoundsUntilPossibleActivation + "\n";
    }

}

File: Project3_Python/Animals/Human.py
from Organism import OrganismsEnum, Point
from Animal import Animal
from enum import Enum


class Human(Animal):
    __DEFAULT_INITIATIVE = 4
    __NUMBER_OF_TURNS_WITH_ABILITY_ON = 5

    class Move(Enum):
        NOMOVE = 0
        LEFT = 1
        RIGHT = 2
        UP = 3
        DOWN = 4

    def __init__(self, world, position, strength=None, def_strength=5, is_ability_on=False, n_until_activation=0):
        super().__init__(world, position)
        self.type_of_organism = OrganismsEnum.Human
        self._initiative = Human.__DEFAULT_INITIATIVE
        self.DEFAULT_STRENGTH = def_strength
        if strength is not None:
            self.strength = strength
        else:
            self.strength = self.DEFAULT_STRENGTH
        self.is_ability_on = is_ability_on
        self.number_of_rounds_until_possible_activation = n_until_activation

        self.current_move = Human.Move.NOMOVE

    def special_ability(self):
        if self.number_of_rounds_until_possible_activation == 0 and not self.is_ability_on:
            self.DEFAULT_STRENGTH = self.strength
            self.strength += Human.__NUMBER_OF_TURNS_WITH_ABILITY_ON
            self.is_ability_on = True
            message = "Special Ability Activated. Current strength: " + str(self.strength) + \
                      ".\nWill decrease until strength is back to normal."
        elif self.number_of_rounds_until_possible_activation != 0:
            message = "Cannot activate Special Ability.\nWill be possible in: " + \
                      str(self.number_of_rounds_until_possible_activation) + " turns."
        else:
            message = "Ability is already on. Current strength: " + str(self.strength) + "."
        return message

    def decrease_ability(self):
        if self.is_ability_on:
            self.strength -= 1
            if self.strength <= self.DEFAULT_STRENGTH:
                self.number_of_rounds_until_possible_activation = 5
                self.is_ability_on = False

    def decrease_ability_counter(self):
        if self.number_of_rounds_until_possible_activation != 0:
            self.number_of_rounds_until_possible_activation -= 1

    def handle_special_ability(self):
        self.decrease_ability()
        self.decrease_ability_counter()

    def change_move(self, new_move):
        self.current_move = new_move

    def get_new_human_position(self):
        if self.current_move == Human.Move.LEFT and self._position.x >= self._step:
            new_position = Point(self._position.x - self._step, self._position.y)
        elif self.current_move == Human.Move.RIGHT and self._position.x < self._world.get_size_of_world().x - self._step:
            new_position = Point(self._position.x + self._step, self._position.y)
        elif self.current_move == Human.Move.UP and self._position.y >= self._step:
            new_position = Point(self._position.x, self._position.y - self._step)
        elif self.current_move == Human.Move.DOWN and self._position.y < self._world.get_size_of_world().y - self._step:
            new_position = Point(self._position.x, self._position.y + self._step)
        else:
            return Point(-1, -1)
        return new_position

    def action(self):
        new_position = self.get_new_human_position()
        if new_position.x == -1:
            return False
        self._prev_position = self._position
        self._position = new_position
        self.try_to_go_to_position()
        return True

    def __str__(self):
        return "Human"

    def save_organism(self):
        return str(self.type_of_organism.value) + " " + str(self._position.x) + " " + str(self._position.y) + " " + \
               str(self._prev_position.x) + " " + str(self._prev_position.y) + " " + str(self.strength) + " " + \
               str(self.DEFAULT_STRENGTH) + " " + str(self.is_ability_on) + " " + \
               str(self.number_of_rounds_until_possible_activation) + "\n"
